src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,494 @@
+            .orShould()
+            .beAnnotatedWith(GlobalStatefulCheck.class)
+            .orShould(beSuppressedClass);
+
+        classesWithMutableFieldsShouldBeStateful.check(CHECKSTYLE_CHECKS);
+    }
+
+    /**
+     * ArchCondition checking fields are immutable.
+     */
+    private static final class ImmutableFieldArchCondition extends ArchCondition<JavaField> {
+        private ImmutableFieldArchCondition() {
+            super("be among immutable types");
+        }
+
+        /**
+         * Whether a field is immutable or not. We can't check complete immutability, reference
+         * immutability can't be achieved. According to this method a field is immutable if it
+         * fulfills any one of these conditions:
+         * <ol>
+         *     <li>If raw type of a field is an enum.</li>
+         *     <li>If the field is a zero size array. We have manually created a list of zero
+         *     size arrays which is checked for containing that particular field.</li>
+         *     <li>If the field is of a primitive type or one among the
+         *     {@link ImmutabilityTest#IMMUTABLE_TYPES}.</li>
+         *     <li>If the field is a parameterized field, it is checked that all the parameterized
+         *     types are among {@link ImmutabilityTest#IMMUTABLE_TYPES}.</li>
+         * </ol>
+         *
+         * @param javaField field to check
+         * @return {@code true} if field is immutable
+         */
+        private static boolean isImmutable(JavaField javaField) {
+            final JavaClass rawType = javaField.getRawType();
+            boolean isImmutable = true;
+
+            final boolean isEnumConstantOrEmptyArray = rawType.isEnum()
+                || ZERO_SIZE_ARRAY_FIELDS.contains(javaField.getFullName());
+
+            if (!isEnumConstantOrEmptyArray) {
+                final String rawTypeName = rawType.getName();
+                final boolean isImmutableType = !PRIMITIVE_TYPES.contains(rawTypeName)
+                    && !IMMUTABLE_TYPES.contains(rawTypeName);
+
+                if (isImmutableType) {
+                    final JavaType javaType = javaField.getType();
+                    isImmutable = false;
+                    if (javaType instanceof JavaParameterizedType) {
+                        final JavaParameterizedType parameterizedType =
+                            (JavaParameterizedType) javaType;
+                        isImmutable = parameterizedType.getActualTypeArguments().stream()
+                            .allMatch(actualTypeArgument -> {
+                                return IMMUTABLE_TYPES.contains(
+                                    actualTypeArgument.toErasure().getName());
+                            });
+                    }
+                }
+            }
+            return isImmutable;
+        }
+
+        @Override
+        public void check(JavaField item, ConditionEvents events) {
+            if (!isImmutable(item)) {
```

Comment:
Wouldn't this be better and easier to ready to split into like 3 different methods?
`!isRawTypeImmutable || !isEnumConstantOrEmptyArray || !isParameterizedTypeImmutable`

Reply:
The Javadoc mentions the conditions `isImmutable(..)` checks, but if you insist, I can modify the method.

Reply:
I insist. The javadoc comments can be a bit simpler now.

Reply:
Done.